### ISSUE TITLE
are jupyter noteboos safe to mark trusted?

### ISSUE THREAD

**Issue by 1rdsk1** [jupyter/help#330]
Hi
I understand that ipynb could have malicious code.  How can I tell if a notebook is sake to use please?

**Comment by takluyver** [comment]
Notebooks can contain two kinds of code:

1. Code in code cells. This you just have to look at before you run it and decide if what it's doing looks reasonable. It doesn't do anything until you explicitly run those cells, so it's not a worry when loading the notebook.

2. Javascript code in the output. This is invisible when you view the notebook, but it can talk to the kernel and send it Python code (or whatever language the kernel runs) to execute on your computer. This is what the trust mechanism deals with.

When you load an *untrusted* notebook, it won't display any output that can contain Javascript (or CSS). This is meant to make it safe to open a notebook to look at. If you run all the cells in a notebook, the output has all been generated on your computer from code you can see, so the notebook becomes trusted. If you're sure of where it comes from, you can also bypass this by marking it trusted directly. A trusted notebook can display HTML/Javascript/CSS when you open it, so it could immediately start running code on your computer.

You could open the ipynb file in a text editor to manually inspect outputs, but it's not very easy to read. Another option is to use [nbstripout](https://pypi.python.org/pypi/nbstripout) to get rid of all outputs, and then open a notebook with only the code + markdown.

TL;DR - don't mark notebooks as trusted unless you're really sure that you can trust whoever made them. It's like downloading and running a program without looking at it. And like everything on the internet, remember that someone might not be who they claim to be.

**Comment by rsvp** [comment]
@takluyver hi Thomas, thanks for the clear explanation.
Now quoted at https://github.com/rsvp/fecon235/wiki/Security:-trust

**Comment by 1rdsk1** [comment]
Hi Thomas, many thanks for the helpful info.  After reading this I thought that if I download an open a notebook it should be marked as untrusted.  However, I just did that and was surprised to see that the notebook opened as trusted.  Are there any rules on which notebooks get trusted and which dont please?

**Comment by takluyver** [comment]
> After reading this I thought that if I download an open a notebook it should be marked as untrusted. However, I just did that and was surprised to see that the notebook opened as trusted. Are there any rules on which notebooks get trusted and which dont please?

There are a couple of ways that can happen:

1. A notebook with no code cells when you open it will always be trusted. This is safe because it should only be possible to put JS/CSS in the output area attached to a code cell. Markdown cells are sanitised even for trusted notebooks, so they're safe.
2. The trust is based on signing the notebook content. So if you've had an identical notebook at some point which was trusted, another copy of the file will also be trusted even if it's a fresh download.

If you think neither of these should apply, please tell us on security@ipython.org so we can investigate.

> Also is nbstripout the same as cell-all output-clear within Jupyter? Is it ok to open the ipynb in Jupyter and just clear the output before marking it as trusted? Would that remove any risk of malicious code?

Yep. In fact, if you clear all output and save, that should directly make the notebook trusted.
